include/freerdp/utils/pod_arrays.h

```diff
@@ -0,0 +1,146 @@
+ */
+#ifndef FREERDP_UTILS_POD_ARRAYS_H_
+#define FREERDP_UTILS_POD_ARRAYS_H_
+
+#include <winpr/wtypes.h>
+#include <freerdp/freerdp.h>
+
+#define POD_ARRAYS_DECL(T, TLOWER) \
+	typedef struct \
+	{ \
+		T* values; \
+		size_t nvalues; \
+	} Array##T; \
+	typedef BOOL Array##T##Cb(T v, void* data); \
+	\
+	FREERDP_API void array_##TLOWER##_init(Array##T* a); \
+	FREERDP_API size_t array_##TLOWER##_size(const Array##T* a); \
+	FREERDP_API T* array_##TLOWER##_data(const Array##T* a); \
+	FREERDP_API const T* array_##TLOWER##_cdata(const Array##T* a); \
+	FREERDP_API T array_##TLOWER##_get(const Array##T* a, size_t idx); \
+	FREERDP_API void array_##TLOWER##_set(Array##T* a, size_t idx, T v); \
+	FREERDP_API BOOL array_##TLOWER##_append(Array##T* a, T v); \
+	FREERDP_API BOOL array_##TLOWER##_contains(const Array##T* a, T v); \
+	FREERDP_API BOOL array_##TLOWER##_foreach(const Array##T* a, Array##T##Cb cb, void* data); \
+	FREERDP_API void array_##TLOWER##_reset(Array##T* a); \
+	FREERDP_API void array_##TLOWER##_uninit(Array##T* a)
+
+#define POD_ARRAYS_IMPL(T, TLOWER) \
+	void array_##TLOWER##_init(Array##T* a) \
+	{ \
+		WINPR_ASSERT(a); \
+		a->values = NULL; \
+		a->nvalues = 0; \
+	} \
+	\
+	size_t array_##TLOWER##_size(const Array##T* a) \
+	{ \
+		WINPR_ASSERT(a); \
+		return a->nvalues; \
+	} \
+	\
+	T* array_##TLOWER##_data(const Array##T* a) \
+	{ \
+		WINPR_ASSERT(a); \
+		return a->values; \
+	} \
+	\
+	const T* array_##TLOWER##_cdata(const Array##T* a) \
+	{ \
+		WINPR_ASSERT(a); \
+		return (const T*)a->values; \
+	} \
+	\
+	T array_##TLOWER##_get(const Array##T* a, size_t idx) \
+	{ \
+		WINPR_ASSERT(a); \
+		WINPR_ASSERT(a->nvalues > idx); \
+		return a->values[idx]; \
+	} \
+	\
+	void array_##TLOWER##_set(Array##T* a, size_t idx, T v) \
+	{ \
+		WINPR_ASSERT(a); \
+		WINPR_ASSERT(a->nvalues > idx); \
+		a->values[idx] = v; \
+	} \
+	\
+	BOOL array_##TLOWER##_append(Array##T* a, T v) \
+	{ \
+		WINPR_ASSERT(a); \
+		T* tmp = realloc(a->values, sizeof(T) * (a->nvalues + 1)); \
+		if (!tmp) \
+			return FALSE; \
+		\
+		tmp[a->nvalues] = v; \
+		a->values = tmp; \
+		a->nvalues++; \
+		return TRUE; \
+	} \
+	\
+	BOOL array_##TLOWER##_contains(const Array##T* a, T v) \
+	{ \
+		WINPR_ASSERT(a); \
+		UINT32 i; \
+		\
+		for (i = 0; i < a->nvalues; i++) \
+		{ \
+			if (a->values[i] == v) \
+				return TRUE; \
+		} \
+		\
+		return FALSE; \
+	} \
+	\
+	BOOL array_##TLOWER##_foreach(const Array##T* a, Array##T##Cb cb, void* data) \
+	{ \
+		WINPR_ASSERT(a); \
+		size_t i; \
+		for (i = 0; i < a->nvalues; i++) \
+		{ \
+			if (!cb(a->values[i], data)) \
+				return FALSE; \
+		} \
+		\
+		return TRUE; \
+	} \
+	\
+	void array_##TLOWER##_reset(Array##T* a) \
+	{ \
+		WINPR_ASSERT(a); \
+		a->nvalues = 0; \
+	} \
+	\
+	void array_##TLOWER##_uninit(Array##T* a) \
+	{ \
+		WINPR_ASSERT(a); \
+		free(a->values); \
+		\
+		a->values = NULL; \
+		a->nvalues = 0; \
+	}
+
+
+POD_ARRAYS_DECL(UINT16, uint16);
+POD_ARRAYS_DECL(UINT32, uint32);
```

Comment:
why only unsigned types, also why no `UINT64`?

Reply:
Sure we can add as much as needed, but I'd rather add a new one as we make use of it.